### PR TITLE
H7 cache

### PR DESCRIPTION
**Pull request by antisvin**
Not sure about HAL_NVIC_SetPriority(DMA1_Stream0_IRQn, 0, 0); changing priority from 5 to 0. This is for ADC DMA. It changed after re-exporting from Cube and I don't think that the old value was reasonable, so I've left the new one.

### CODE REVIEW

File: Genius/Src/stm32h7xx_it.c [hunk 1]
```diff
@@ -204,6 +235,35 @@ void EXTI15_10_IRQHandler(void)
...
+  if(huart->Instance->ISR & UART_FLAG_IDLE){
+    /* This part is important */
+    /* Clear IDLE flag by reading status and data registers */
+    //volatile uint32_t tmp;                  /* Must be volatile to prevent optimizations */
+    //tmp = huart->Instance->ISR;              /* Read status register */
+    //tmp = huart->Instance->RDR;              /* Read data register */
+    //(void)tmp;                              /* Prevent compiler warnings */
+    __HAL_UART_CLEAR_IDLEFLAG(huart);
+    if(huart->hdmarx != NULL)
+      ((DMA_Stream_TypeDef*)(huart->hdmarx->Instance))->CR &= ~DMA_SxCR_EN;       /* Disabling DMA will force transfer complete interrupt if enabled */
```
Comment (pingdynasty) [inline Genius/Src/stm32h7xx_it.c:260]: can probably improve this with
```__HAL_DMA_DISABLE(huart->hdmarx)```

File: Source/sdram.c [hunk 2]
```diff
@@ -26,6 +26,7 @@ void MPU_Config(void){
   /* Disable the MPU */
   HAL_MPU_Disable();
 
+#ifndef STM32H743xx
```
Comment (pingdynasty) [inline Source/sdram.c:29]: pls change to positive logic ie 
#ifdef STM32H743xx

File: Source/uart.cpp [hunk 3]
```diff
@@ -2,12 +2,16 @@
 #include "device.h"
 #include "errorhandlers.h"
 #include "SerialBuffer.hpp"
+#ifdef STM32H743xx
+#include "stm32h7xx_hal.h"
+#else
 #include "stm32f4xx_hal.h"
+#endif
```
Comment (pingdynasty) [inline Source/uart.cpp:9]: should not be necessary when #including device.h, which includes hardware.h which includes device specific hal.h

File: Source/uart.cpp [hunk 4]
```diff
@@ -58,7 +62,7 @@ extern "C"{
 
   void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart){
     // what is the correct size if IDLE interrupts?
-    size_t size = huart->RxXferSize - huart->hdmarx->Instance->NDTR;
+    size_t size = huart->RxXferSize -  ((DMA_Stream_TypeDef*)huart->hdmarx->Instance)->NDTR;
```
Comment (pingdynasty) [inline Source/uart.cpp:65]: there's a `__HAL_DMA_GET_COUNTER()` macro you could use

**Review by pingdynasty — CHANGES_REQUESTED** [review]

**Comment by pingdynasty** [comment]
nice work

**Comment by antisvin** [comment]
But nice work with working MIDI out > nice work :-/ Anyway, will also try to enable it for Magus, could be a nice addition for next FW release - and the only EXT out option that we can fully support.

**Pull request merged**